### DISCUSSION TITLE
Own Bootloader

### DISCUSSION THREAD

**Discussion by ghost** [phil-opp/blog_os#944] [Questions] [answered]
Welcome back. I have a bootloader written in assembly language. And I wonder how to connect this to the kernel. I was browsing the files in the target folder and came across a file called "blog_os" in the path: ./target/x86_64-blog_os/debug/blog_os. I also noticed that every time qemu starts up, it starts the botimage-blog_os file. The bootimage-blog_os file has a bootloader provided because it is bootable. If I'm right when I use cargo objcopy to create blog_os.bin from blog_os and attach my bootloader to it, it should work. I'd like to know what you @phil-opp  think.

**Comment by ghost** [comment]
@phil-opp  Recently you advised me to connect my bootloader to the system I have to compile it as static lib when I do this then you can't do anything about it. Neither use cargo objcopy nor anything else.

**Comment by phil-opp** [comment, marked as answer]
It's difficult to answer your question in general since it depends on your bootloader implementation and your build environment (e.g. Windows or Linux). You normally have two options:

1. Link your kernel directly with the bootloader by compiling at as a staticlib.

    This requires that you use the same file types for the bootloader and kernel. If you're using Linux, you can just link the resulting `.a` staticlib with your other ELF object files. If you're using Windows, you probably need to compile your kernel as a Windows-compatible staticlib by modifying your target JSON file. I've never tried this, so I can't help you with it unfortunately.
2. Build the kernel as a separate executable and load it in your bootloader.

    This is the approach that we use on the blog, but it's a bit more difficult to implement. You need to include the kernel executable in your disk image somehow, e.g. by using [`include_bytes`](https://doc.rust-lang.org/stable/core/macro.include_bytes.html), by including it in your linker script, or by implementing a filesystem.

    The second step is then to parse the kernel executable in your bootloader and set up the execution environment accordingly (e.g. set up page table mappings). This step depends on the executable format that your kernel uses. For example, on the blog we use the [ELF](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) format, which you can for example parse with the [`xmas_elf`](https://docs.rs/xmas-elf/0.7.0/xmas_elf/) crate (there are many other ELF parsing crates).

If you're interested in details, I recommend reading the code of the `bootloader` crate.

**Reply by ghost** [reply]
I looked at the Bootloader code and how to set KERNEL and KERNEL_MANIFEST variables in Windows ?

**Reply by phil-opp** [reply]
These are normal environment variables. The former should point to your kernel execuable (typically in your `target` folder) and the latter to the `Cargo.toml` of the kernel

**Reply by ghost** [reply]
So I have to put it in the system environment variables?

**Reply by phil-opp** [reply]
No. You can e.g. use the `set` command to set it temporily (https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/set_1)

**Reply by ghost** [reply]
You wrote that on Linux it is easy to combine a library with an assembler. What if I had used WSL, it might have worked. What commands to use when it comes to Linux

**Reply by phil-opp** [reply]
You can link the `.a` file like a normal `.o` file by passing it to your linker.

**Reply by phil-opp** [reply]
See https://os.phil-opp.com/set-up-rust/#integrating-rust

**Comment by ghost** [comment]
@phil-opp What's wrong with this command 
```bash
ld -n --gc-sections -T /src/arch/x86_64/linker.ld -o kernel.bin \
  /build/arch/x86_64/boot.o \
  /build/arch/x86_64/long_mode_init.o \
  /build/arch/x86_64/multiboot_header.o \
  /target/x86_64-blog_os/debug/libblog_os.a 
```
I mean the path to linker.ld
